src/qmoperators/two_electron/ExchangePotential.cpp

```diff
@@ -171,47 +170,48 @@ void ExchangePotential::calcExchange_kij(double prec, Orbital phi_k, Orbital phi
     // if the product is smaller than the target precision,
     // the result is expected to be negligible
     Timer timer_ij;
-    Orbital rho_ij = phi_i.paramCopy();
-    mrcpp::cplxfunc::multiply(rho_ij, phi_i.dagger(), phi_j, prec_m1, true, true);
+    Orbital rho_ij = phi_i.paramCopy(true);
+    mrcpp::multiply(rho_ij, phi_i, phi_j, prec_m1, true, true, true);
     timer_ij.stop();
     if (rho_ij.norm() < prec) return;
 
-    auto N_i = phi_i.getNNodes(NUMBER::Total);
-    auto N_j = phi_j.getNNodes(NUMBER::Total);
-    auto N_ij = rho_ij.getNNodes(NUMBER::Total);
+    auto N_i = phi_i.getNNodes();
+    auto N_j = phi_j.getNNodes();
+    auto N_ij = rho_ij.getNNodes();
     auto norm_ij = rho_ij.norm();
+    // For now we assume all phi are complex or all ar real.
 
-    // prepare vector used to steer precision of Poisson application
-    mrcpp::FunctionTreeVector<3> phi_opt_vec;
-    if (phi_k.hasReal()) phi_opt_vec.push_back(std::make_tuple(1.0, &phi_k.real()));
-    if (phi_k.hasImag()) phi_opt_vec.push_back(std::make_tuple(1.0, &phi_k.imag()));
-
-    if (phi_j.hasReal() and &phi_j != &phi_k) phi_opt_vec.push_back(std::make_tuple(1.0, &phi_j.real()));
-    if (phi_j.hasImag() and &phi_j != &phi_k) phi_opt_vec.push_back(std::make_tuple(1.0, &phi_j.imag()));
-
-    if (phi_i.hasReal() and &phi_i != &phi_k and &phi_i != &phi_j) phi_opt_vec.push_back(std::make_tuple(1.0, &phi_i.real()));
-    if (phi_i.hasImag() and &phi_i != &phi_k and &phi_i != &phi_j) phi_opt_vec.push_back(std::make_tuple(1.0, &phi_i.imag()));
+    bool RealOrbitals = phi_i.isreal();
 
+    // prepare vector used to steer precision of Poisson application
+    mrcpp::FunctionTreeVector<3, double> phi_opt_vec_real;
+    mrcpp::FunctionTreeVector<3, ComplexDouble> phi_opt_vec_cplx;
+    if (RealOrbitals) {
+        if (phi_k.isreal() and phi_k.getNNodes() > 0) phi_opt_vec_real.push_back(std::make_tuple(1.0, phi_k.CompD[0]));
```

Comment:
it should be possible to do this instead. With `emplace_back` the compiler should avoid temporaries:
```suggestion
        if (phi_k.isreal() and phi_k.getNNodes() > 0) phi_opt_vec_real.emplace_back(1.0, phi_k.CompD[0]);
```